### PR TITLE
chore: implemented property simplification

### PR DESCRIPTION
**Pull request by jonaslagoni**
**Description**
This PR introduces property simplification together with docs and tests.

**Related issue(s)**
Related to https://github.com/asyncapi/shape-up-process/issues/45

### CODE REVIEW

File: src/models/CommonModel.ts [hunk 1]
```diff
@@ -25,4 +25,74 @@ export class CommonModel extends CommonSchema<CommonModel>{
...
+    /**
+     * Only merge if left side is undefined and right side is sat OR both sides are defined
+     * 
+     * @param mergeTo CommonModel to merge into
+     * @param mergeFrom CommonModel to merge values from
+     */
+    static mergeCommonModels(mergeTo: CommonModel, mergeFrom: CommonModel, originalSchema: Schema) : CommonModel {
+        const mergeToProperties = mergeTo.properties;
+        const mergeFromProperties = mergeFrom.properties;
+        if(mergeToProperties === undefined && mergeFromProperties !== undefined){
```
Comment (magicmatatjahu) [inline src/models/CommonModel.ts:39]: I think that logic like:

```ts
if (mergeFromProperties !== undefined) {
  if (mergeToProperties === undefined) {
     mergeTo.properties = mergeFromProperties;
  } else {
     ...
  }
}
```

is easier to read :) But it's only a suggestion.

File: src/simplification/SimplifyProperties.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ */
+export default function simplifyProperties(schema: Schema) : output {
+  let models : CommonModel[] | undefined;
+  let commonProperties : { [key: string]: CommonModel; } | undefined;
+  const addToModels = (model : CommonModel[]) => {
+    if(models === undefined){
+      models = [];
+    }
+    models = [...models, ...model];
+  }
```
Comment (magicmatatjahu) [inline src/simplification/SimplifyProperties.ts:19]:
```ts
const addToModels = (model : CommonModel[] = []) => { models = [...(models || []), ...model]; }
```

**Review by magicmatatjahu — CHANGES_REQUESTED** [review]
Great! As we discussed, we should go in next week with another handling of CommonModel to handle complex cases like unions, traits etc.

File: src/simplification/SimplifyProperties.ts (hunk 2 above)
Comment (jonaslagoni) [reply src/simplification/SimplifyProperties.ts:19]: Smart 🧠

File: src/models/CommonModel.ts (hunk 1 above)
Comment (jonaslagoni) [reply src/models/CommonModel.ts:39]: Yea I agree :)

File: src/models/CommonModel.ts [hunk 3]
```diff
@@ -25,4 +25,76 @@ export class CommonModel extends CommonSchema<CommonModel>{
...
+                    if(mergeToProperties![propName] !== undefined){
+                      mergeTo.properties![propName] = CommonModel.mergeCommonModels(mergeToProperties![propName], prop, originalSchema);
+                    } else {
+                      mergeTo.properties![propName] = prop;
+                    }
+                });
+            }
+        }
+
+        if(mergeTo.items === undefined && mergeFrom.items !== undefined){
```
Comment (magicmatatjahu) [inline src/models/CommonModel.ts:53]: I forgot to mention that if you apply changes, you can add them here too, I mean changed if-s with undefined

Comment (jonaslagoni) [reply src/models/CommonModel.ts:53]: ahh! right, 2 sec

**Review by magicmatatjahu — APPROVED** [review]

**Pull request merged**